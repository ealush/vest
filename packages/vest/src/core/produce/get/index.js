import collectFailureMessages from '../collectFailureMessages';

/**
 * @param {'errors'|'warnings'} severityKey lookup severity
 * @param {string} [fieldName]
 * @returns suite or field's errors or warnings.
 */
const get = (severityKey, fieldName) => {
  const res = collectFailureMessages(severityKey, { fieldName });

  if (fieldName) {
    return res[fieldName] || [];
  } else {
    return res;
  }
};

export default get;
