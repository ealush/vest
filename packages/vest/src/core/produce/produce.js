import createCache from 'cache';
import context from 'ctx';
import genTestsSummary from 'genTestsSummary';
import getFailures from 'getFailures';
import getFailuresByGroup from 'getFailuresByGroup';
import hasFailures from 'hasFailures';
import hasFailuresByGroup from 'hasFailuresByGroup';
import hasRemainingTests from 'hasRemainingTests';
import isFunction from 'isFunction';
import { SEVERITY_GROUP_ERROR, SEVERITY_GROUP_WARN } from 'resultKeys';
import { HAS_WARNINGS, HAS_ERRORS } from 'sharedKeys';
import {
  useTestCallbacks,
  useTestObjects,
  useOptionalFields,
} from 'stateHooks';
import withArgs from 'withArgs';

const cache = createCache(20);

/**
 * @param {boolean} [isDraft]
 * @returns Vest output object.
 */
const produce = isDraft => {
  const { stateRef, optional } = context.use();
  const [testObjects] = useTestObjects();

  const ctxRef = { stateRef, optional };

  return cache(
    [testObjects, isDraft],
    context.bind(ctxRef, () =>
      [
        [HAS_ERRORS, hasFailures, SEVERITY_GROUP_ERROR],
        [HAS_WARNINGS, hasFailures, SEVERITY_GROUP_WARN],
        ['getErrors', getFailures, SEVERITY_GROUP_ERROR],
        ['getWarnings', getFailures, SEVERITY_GROUP_WARN],
        ['hasErrorsByGroup', hasFailuresByGroup, SEVERITY_GROUP_ERROR],
        ['hasWarningsByGroup', hasFailuresByGroup, SEVERITY_GROUP_WARN],
        ['getErrorsByGroup', getFailuresByGroup, SEVERITY_GROUP_ERROR],
        ['getWarningsByGroup', getFailuresByGroup, SEVERITY_GROUP_WARN],
      ]
        .concat(
          [['isValid', isValid]],
          isDraft ? [] : [['done', withArgs(done)]]
        )
        .reduce((properties, [name, fn, severityKey]) => {
          properties[name] = context.bind(ctxRef, fn, severityKey);
          return properties;
        }, genTestsSummary())
    )
  );
};

export default produce;

/**
 * Registers done callbacks.
 * @param {string} [fieldName]
 * @param {Function} doneCallback
 * @register {Object} Vest output object.
 */
function done(args) {
  const [callback, fieldName] = args.reverse();
  const { stateRef } = context.use();

  const output = produce();

  // If we do not have any test runs for the current field
  const shouldSkipRegistration =
    fieldName &&
    (!output.tests[fieldName] || output.tests[fieldName].testCount === 0);

  if (!isFunction(callback) || shouldSkipRegistration) {
    return output;
  }

  const cb = context.bind({ stateRef }, () =>
    callback(produce(/*isDraft:*/ true))
  );

  // is suite finished || field name exists, and test is finished
  const shouldRunCallback =
    !hasRemainingTests() || (fieldName && !hasRemainingTests(fieldName));

  if (shouldRunCallback) {
    cb();
    return output;
  }

  const [, setTestCallbacks] = useTestCallbacks();
  setTestCallbacks(current => {
    if (fieldName) {
      current.fieldCallbacks[fieldName] = (
        current.fieldCallbacks[fieldName] || []
      ).concat(cb);
    } else {
      current.doneCallbacks.push(cb);
    }
    return current;
  });

  return output;
}

function isValid() {
  const result = produce();

  if (result.hasErrors()) {
    return false;
  }

  const [testObjects] = useTestObjects();

  if (testObjects.length === 0) {
    return false;
  }

  const [optionalFields] = useOptionalFields();

  for (const test in result.tests) {
    if (!optionalFields[test] && result.tests[test].testCount === 0) {
      return false;
    }
  }

  return true;
}
