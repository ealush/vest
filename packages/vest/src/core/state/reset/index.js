import { setSuites } from '..';
import throwError from '../../../lib/throwError';
import getSuiteState from '../getSuiteState';
import setCanceled from '../setCanceled';

/**
 * Cleans up a suite from state.
 * @param {string} suiteId
 */
const reset = suiteId => {
  if (!suiteId) {
    throwError('`vest.reset` must be called with suiteId.');
  }

  const suite = getSuiteState(suiteId);

  if (!suite) {
    return;
  }

  setCanceled(...(suite.pending || []));
  setCanceled(...(suite.lagging || []));

  setSuites(state => {
    delete state[suiteId];
    return state;
  });
};

export default reset;
