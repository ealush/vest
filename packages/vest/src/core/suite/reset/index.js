import { OPERATION_MODE_STATEFUL } from '../../../constants';
import throwError from '../../../lib/throwError';
import context from '../../context';
import getState from '../getState';
import register from '../register';
import remove from '../remove';

/**
 * Resets suite to its initial state
 * @param {String} suiteId
 */
const reset = suiteId => {
  if (!suiteId) {
    throwError('`vest.reset` must be called with suiteId.');
  }

  let name = suiteId;
  try {
    name = getState(suiteId).name;
    remove(suiteId);
  } catch {
    /* */
  }

  context.run(
    { name, suiteId, operationMode: OPERATION_MODE_STATEFUL },
    register
  );
};

export default reset;
