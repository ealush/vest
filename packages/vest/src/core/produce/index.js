import createCache from '../../lib/cache';
import copy from '../../lib/copy';
import hasRemainingTests from '../suite/hasRemainingTests';
import {
  SEVERITY_GROUP_ERROR,
  SEVERITY_GROUP_WARN,
} from '../test/lib/VestTest/constants';
import genTestsSummary, { countFailures } from './genTestsSummary';
import get from './get';
import getByGroup from './getByGroup';
import has from './has';
import hasByGroup from './hasByGroup';

const cache = createCache(20);

/**
 * Registers done callbacks.
 * @param {Object} state
 * @param {string} [fieldName]
 * @param {Function} doneCallback
 * @register {Object} Vest output object.
 */
const done = (stateRef, ...args) => {
  const { length, [length - 1]: callback, [length - 2]: fieldName } = args;

  const output = produce(stateRef);

  // If we do not have any tests for current field
  const shouldSkipRegistration = fieldName && !output.tests[fieldName];

  if (typeof callback !== 'function' || shouldSkipRegistration) {
    return output;
  }

  const state = stateRef.current();

  const cb = () => callback(produce(stateRef, { draft: true }));
  const isFinishedTest = fieldName && !hasRemainingTests(state, fieldName);
  const isSuiteFinished = !hasRemainingTests(state);
  const shouldRunCallback = isFinishedTest || isSuiteFinished;
  if (shouldRunCallback) {
    cb();
    return output;
  }

  stateRef.patch(state => {
    if (fieldName) {
      state.fieldCallbacks[fieldName] = [].concat(
        ...(state.fieldCallbacks[fieldName] || []),
        cb
      );
    } else {
      state.doneCallbacks.push(cb);
    }
    return state;
  });

  return output;
};

/**
 * @returns {Object} with only public properties.
 */
const extract = ({ groups, tests, name }) => ({
  groups,
  tests,
  name,
});

/**
 * @param {Object} state
 * @param {Object} Options
 * @param {boolean} [Options.draft]
 * @returns Vest output object.
 */

const produce = (stateRef, { draft } = {}) => {
  const state = stateRef.current();
  return cache(
    [state, draft],
    () =>
      state
      |> copy
      |> genTestsSummary
      |> extract
      |> countFailures
      |> (transformedState =>
        Object.defineProperties(
          transformedState,
          [
            ['hasErrors', has.bind(null, state, SEVERITY_GROUP_ERROR)],
            ['hasWarnings', has.bind(null, state, SEVERITY_GROUP_WARN)],
            ['getErrors', get.bind(null, state, SEVERITY_GROUP_ERROR)],
            ['getWarnings', get.bind(null, state, SEVERITY_GROUP_WARN)],
            [
              'hasErrorsByGroup',
              hasByGroup.bind(null, state, SEVERITY_GROUP_ERROR),
            ],
            [
              'hasWarningsByGroup',
              hasByGroup.bind(null, state, SEVERITY_GROUP_WARN),
            ],
            [
              'getErrorsByGroup',
              getByGroup.bind(null, state, SEVERITY_GROUP_ERROR),
            ],
            [
              'getWarningsByGroup',
              getByGroup.bind(null, state, SEVERITY_GROUP_WARN),
            ],
          ]
            .concat(draft ? [] : [['done', done.bind(null, stateRef)]])
            .reduce((properties, [name, value]) => {
              properties[name] = {
                configurable: true,
                enumerable: true,
                value,
                writeable: true,
              };
              return properties;
            }, {})
        ))
  );
};

export default produce;
