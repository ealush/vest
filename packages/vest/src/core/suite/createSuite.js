import createStateRef from 'createStateRef';

import asArray from 'asArray';
import context from 'ctx';
import genId from 'genId';
import isFunction from 'isFunction';
import mergeExcludedTests from 'mergeExcludedTests';
import produce from 'produce';
import createState from 'state';
import throwError from 'throwError';
import usePending from 'usePending';
import useTestObjects from 'useTestObjects';
import withArgs from 'withArgs';
/**
 * Initializes a validation suite, creates a validation context.
 * @param {String} [name]     Identifier for validation suite.
 * @param {Function} tests  Validation suite body.
 * @returns {Function} validator function.
 */
const createSuite = withArgs(args => {
  const [tests, name] = args.reverse();

  if (!isFunction(tests)) {
    throwError(
      'Suite initialization error. Expected `tests` to be a function.'
    );
  }

  const state = createState();

  const handlers = [];
  const stateRef = createStateRef(state, { suiteId: genId(), name });

  /*
    context.bind returns our `validate` function
    We then wrap it with defineProperties to add
    the `get`, and `reset` functions.

  */
  const suite = context.bind({ stateRef }, function () {
    const [previousTestObjects] = useTestObjects();
    const [{ pending }, setPending] = usePending();
    state.reset();

    // Move all the active pending tests to the lagging array
    setPending({ lagging: pending, pending: [] });

    // Run the consumer's callback
    tests.apply(null, arguments);

    // Merge all the skipped tests with their previous results
    mergeExcludedTests(previousTestObjects);

    return produce();
  });
  suite.get = context.bind({ stateRef }, produce, /*isDraft:*/ true);
  suite.reset = state.reset;
  suite.remove = context.bind({ stateRef }, name => {
    const [testObjects] = useTestObjects();

    // We're mutating the array in `cancel`, so we have to first copy it.
    asArray(testObjects).forEach(testObject => {
      if (testObject.fieldName === name) {
        testObject.cancel();
      }
    });
  });

  suite.subscribe = function (handler) {
    if (!isFunction(handler)) return;

    handlers.push(handler);

    handler({
      type: 'suiteSubscribeInit',
      suiteState: stateRef.current(),
    });
  };

  return suite;
});

export default createSuite;
