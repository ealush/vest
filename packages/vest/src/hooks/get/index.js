import produce from '../../core/produce';
import getSuiteState from '../../core/state/getSuiteState';
import throwError from '../../lib/throwError';

/**
 * @param {String} suiteId Suite id to find
 * @returns {Object} Up to date state copy.
 */
const get = suiteId => {
  if (!suiteId) {
    throwError('`get` hook was called without a suite name.');
  }

  const state = getSuiteState(suiteId);
  return produce(state, { draft: true });
};

export default get;
