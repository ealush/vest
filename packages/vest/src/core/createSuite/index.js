import { OPERATION_MODE_STATEFUL } from '../../constants';
import runWithContext from '../../lib/runWithContext';
import singleton from '../../lib/singleton';
import validateSuiteParams from '../../lib/validateSuiteParams';
import produce from '../produce';
import { getSuite } from '../state';
import getSuiteState from '../state/getSuiteState';
import registerSuite from '../state/registerSuite';
import mergeExcludedTests from '../test/lib/mergeExcludedTests';
import runAsyncTest from '../test/runAsyncTest';

/**
 * Initializes a validation suite, creates a validation context.
 * @param {String} name     Identifier for validation suite.
 * @param {Function} tests  Validation suite body.
 * @returns {Function} validator function.
 */
const createSuite = (name, tests) => {
  validateSuiteParams('vest.create', name, tests);

  const ctx = singleton.useContext();

  const ctxRef = {
    suiteId: ctx?.suiteId || name,
    operationMode: ctx?.operationMode || OPERATION_MODE_STATEFUL,
    name,
  };

  /**
   * Initialize empty suite state. This is not required for vest
   * itself, but it is handy to have a default value when using
   * UI frameworks that might try to get the validation results
   * during initial render. This is irrelevant for stateless mode.
   */
  if (
    ctxRef.operationMode === OPERATION_MODE_STATEFUL &&
    !getSuite(ctxRef.suiteId)
  ) {
    runWithContext(ctxRef, registerSuite);
  }

  // returns validator function
  return (...args) => {
    const output = runWithContext(ctxRef, context => {
      registerSuite();
      const { suiteId } = context;
      tests.apply(null, args);
      mergeExcludedTests(suiteId);

      [...getSuiteState(suiteId).pending].forEach(runAsyncTest);
      return produce(getSuiteState(suiteId));
    });
    return output;
  };
};

export default createSuite;
