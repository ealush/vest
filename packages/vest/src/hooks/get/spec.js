import { dummyTest } from '../../../testUtils/testDummy';
import createSuite from '../../core/createSuite';
import produce from '../../core/produce';
import getSuiteState from '../../core/state/getSuiteState';
import reset from '../../core/state/reset';
import get from '.';

const suiteId = 'form-name';

const validation = () =>
  createSuite(suiteId, () => {
    dummyTest.passing('f1', 'msg');
    dummyTest.failing('f2', 'msg');
    dummyTest.failingWarning('f2', 'msg');
  });

describe('hook: vest.get()', () => {
  describe('When called without suite id', () => {
    it('Should throw an error', () => {
      expect(get).toThrow(
        '[Vest]: `get` hook was called without a suite name.'
      );
    });
  });

  describe('When suite id does not exist', () => {
    it('Should throw an error', () => {
      expect(() => get('I do not exist')).toThrow();
    });
  });

  describe('When suite exists', () => {
    let validate;

    beforeEach(() => {
      validate = validation();
    });

    afterEach(() => {
      reset(suiteId);
    });

    it('Should return produced result', () => {
      validate();
      expect(produce(getSuiteState(suiteId), { draft: true })).isDeepCopyOf(
        get(suiteId)
      );
      expect(get(suiteId)).toMatchSnapshot();
    });
  });
});
