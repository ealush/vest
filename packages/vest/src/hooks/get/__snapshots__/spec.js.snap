// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`hook: vest.get() When suite exists Should return produced result 1`] = `
Object {
  "errorCount": 1,
  "getErrors": [Function],
  "getErrorsByGroup": [Function],
  "getWarnings": [Function],
  "getWarningsByGroup": [Function],
  "groups": Object {},
  "hasErrors": [Function],
  "hasErrorsByGroup": [Function],
  "hasWarnings": [Function],
  "hasWarningsByGroup": [Function],
  "name": "form-name",
  "tests": Object {
    "f1": Object {
      "errorCount": 0,
      "testCount": 1,
      "warnCount": 0,
    },
    "f2": Object {
      "errorCount": 1,
      "errors": Array [
        "msg",
      ],
      "testCount": 2,
      "warnCount": 1,
      "warnings": Array [
        "msg",
      ],
    },
  },
  "warnCount": 1,
}
`;
